### PR TITLE
WithMaxSize

### LINKED ISSUE #4092: A way to annotate and enforce a max size of call parameters
#3749 essentially added a max size of the call parameter. It is done in such way that makes the DX miserable as we lost the type information.

We should be able to come up a way that don't sacrifice the type information but can still enforce the max parameter size.

Could be something like 

```rs
struct WithMaxSize<T: Parameter, S: Get<usize>>;
```

and custom implementation of encode/decode and helper to construct inner type with safety in mind.

### PR DESCRIPTION
**Pull request by xlc**
closes #4092
we may want to revert #3749 

TODOs:
- [ ] docs
- [ ] tests
- [ ] revert changes from #3749 either fully or partially

### CODE REVIEW

File: polkadot/xcm/xcm-builder/src/controller.rs [hunk 1]
```diff
@@ -48,6 +48,7 @@ pub trait ExecuteControllerWeightInfo {
 
 parameter_types! {
 	pub const MaxXcmEncodedSize: u32 = xcm::MAX_XCM_ENCODED_SIZE;
+	pub const MaxXcmEncodedSizeUsize: usize = xcm::MAX_XCM_ENCODED_SIZE as usize;
```
Comment (xlc) [inline polkadot/xcm/xcm-builder/src/controller.rs:51]: Not sure what's the best way to handle this

**Comment by xlc** [comment]
Screenshot 2024-04-18 at 10 01 17 PM

File: substrate/frame/support/src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+use frame_support_procedural::{EqNoBound, CloneNoBound, PartialEqNoBound, RuntimeDebugNoBound};
+use scale_info::TypeInfo;
+use sp_core::Get;
+
+use crate::Parameter;
+
+/// TODO: docs
+pub struct InputWithMaxLength<'a, I: Input, S: Get<usize>> {
+	input: &'a mut I,
+	len: usize,
```
Comment (bkchr) [inline substrate/frame/support/src/utils.rs:28]:
```suggestion
	remaining_len: usize,
```

File: substrate/frame/support/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+impl<'a, I: Input, S: Get<usize>> Input for InputWithMaxLength<'a, I, S> {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		let len = self.input.remaining_len()?;
+		Ok(len.map(|l| l.min(self.len)))
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		let max_len = self.len.min(into.len());
```
Comment (bkchr) [inline substrate/frame/support/src/utils.rs:45]: This can directly fail if `into.len() > self.len`

File: substrate/frame/support/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		self.len -= max_len;
+		self.input.read(&mut into[..max_len])?;
+		Ok(())
+	}
+}
+
+/// TODO: docs
+#[derive(Encode, EqNoBound, PartialEqNoBound, CloneNoBound, TypeInfo, RuntimeDebugNoBound)]
```
Comment (bkchr) [inline substrate/frame/support/src/utils.rs:53]: IMO we should forward the `TypeInfo` of the inner type?

**Comment by xlc** [comment]
Please let me know what should I do with the changes in pallet-xcm
1. Only keep the WithMaxSize change in this PR and fix pallet-xcm in next PR.
2. Fix send and execute and keep send_blob and execute_blob.
3. Completely revert #3749 and then fix send and execute

File: substrate/frame/support/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	pub fn new(value: T) -> Result<Self, ()> {
+		if value.encoded_size() <= S::get() {
+			Ok(Self { value, _phantom: Default::default() })
+		} else {
+			Err(())
+		}
+	}
+
+	pub fn value(self) -> T {
+		self.value
```
Comment (ggwpez) [inline substrate/frame/support/src/utils.rs:73]: Could make sense to impl the `Deref` trait and mark with `WrapperTypeEncode` and `WrapperTypeDecode`.

**Comment by xlc** [comment]
I guess @franciscoaguirre you are the right person to ping?

**Comment by acatangiu** [comment]
@xlc let me verify this fully fixes the issue please before putting in more effort here, can't promise to do it today, but latest Monday

File: substrate/frame/support/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+impl<'a, I: Input, S: Get<usize>> Input for InputWithMaxLength<'a, I, S> {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		self.input.remaining_len().map(|l| l.map(|l| l.min(self.remaining_len)))
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		if into.len() > self.remaining_len {
+			return Err("Not enough data to fill buffer".into());
+		}
+		self.input.read(into)
```
Comment (bkchr) [inline substrate/frame/support/src/utils.rs:46]:
```suggestion
		let res = self.input.read(into)?;
		self.remaining_len -= into.len();
		Ok(res)
```

We actually need to remove it from the remaining length :D

Comment (xlc) [reply substrate/frame/support/src/utils.rs:46]: I really need some tests XD
and do we need to remove it before the read or it wouldn’t matter?

Comment (bkchr) [reply substrate/frame/support/src/utils.rs:46]: I thought about it. If this returns an error, I would assume that we bubble that up any way and the decoding failed.

**Comment by xlc** [comment]
lazy decode the Call will be a better way

**Pull request closed without merging**